File: go/binarysearch/34_search_range.go
package binarysearch

// searchRange 在排序数组(非递减)中查找元素的第一个和最后一个位置
// 二分
func searchRange(nums []int, target int) []int {
	n := len(nums)

	l, r := 0, n-1
	ans := make([]int, 0)

	for l <= r {
		mid := l + (r-l)>>1

		if nums[mid] == target {

			// 往左查看是否还有等于target 的值, 开始位置
			left := mid
			for left >= 0 && nums[left] == target {
				left--
			}
			// 退出上述循环时，left -1 了
			ans = append(ans, left+1)

			// 结束位置的坐标
			right := mid
			for right < n && nums[right] == target {
				right++
			}
			ans = append(ans, right-1)
			return ans
		}

		// mid 大于target 因为nums 是非递减切片
		if nums[mid] > target {
			r = mid - 1
		} else {
			l = mid + 1
		}

	}
	return []int{-1, -1}
}

File: rust/src/binarysearch/_34_search_range.rs
// searchRange 在排序数组(非递减)中查找元素的第一个和最后一个位置
pub fn search_range2(nums: Vec<i32>, target: i32) -> Vec<i32> {
    let n = nums.len() as i32;

    let (mut l, mut r) = (0, n - 1);

    let mut ans = vec![];

    while l <= r {
        let mid = l + ((r - l) >> 1);

        if nums[mid as usize] == target {
            // 向左找， 找到相等的起始位置

            let mut left = mid as i32;
            while left >= 0 && nums[left as usize] == target {
                left -= 1;
            }
            // 退出上面循环时，left 是满足相等的条件下-1 了
            ans.push(left + 1);

            let mut right = mid;

            while right < n && nums[right as usize] == target {
                right += 1;
            }

            ans.push(right - 1);
            return ans;
        }

        if nums[mid as usize] > target {
            r = mid - 1;
        } else {
            l = mid + 1;
        }
    }
    vec![-1, -1]
}
